### PR TITLE
chore(cli-repl): add internal API documentation MONGOSH-948

### PR DESCRIPTION
**Pull request by addaleax**

### CODE REVIEW

File: packages/cli-repl/src/mongosh-repl.ts [hunk 1]
```diff
@@ -613,34 +742,53 @@ class MongoshNodeRepl implements EvaluationListener {
...
   async close(): Promise<void> {
     const rs = this._runtimeState;
     if (rs) {
       this._runtimeState = null;
       rs.repl.close();
       await rs.internalState.close(true);
-      if (this.output.writableLength > 0) {
-        await once(this.output, 'drain');
-      }
+      await new Promise(resolve => this.output.write('', resolve));
```
Comment (addaleax) [inline packages/cli-repl/src/mongosh-repl.ts:768]: This is a small drive-by fix – `.writableLength > 0` does not imply that `'drain'` will be emitted.

The new state matches what we do in the logging code as well.

**Review by addaleax — COMMENTED** [review]
(Adding @alenakhineika as a reviewer even though she’s OOO, it might still be nice for her to take a look at this :))

**Review by rose-m — APPROVED** [review]
Nice, great to have more docs in there! 🎉

File: packages/cli-repl/src/config-directory.ts [hunk 2]
```diff
@@ -5,12 +5,22 @@ import { EventEmitter } from 'events';
 import { bson } from '@mongosh/service-provider-core';
 const { EJSON } = bson;
 
+/**
+ * A set of paths that are used by the shell, typically located
+ * within the user's home directory.
+ */
 export type ShellHomePaths = {
+  /** A path to data the should be shared between machines if the same user is using them, e.g. config data */
```
Comment (lerouxb) [inline packages/cli-repl/src/config-directory.ts:13]: `A path to data the should` should probably be `A path to data THAT should`?

File: packages/cli-repl/src/log-writer.ts [hunk 3]
```diff
@@ -131,10 +157,14 @@ export class MongoLogWriter extends Writable {
     this._target.end(callback);
   }
 
+  /** Wait until all pending data has been written to the underlying stream. */
   async flush(): Promise<void> {
     await new Promise(resolve => this._target.write('', resolve));
   }
 
+  /**
```
Comment (lerouxb) [inline packages/cli-repl/src/log-writer.ts:165]: I'm trying to infer the style here. Sometimes you do `/** */` on one line and sometimes split over three even when it would have fit on one line. Is there some rule I'm missing? Maybe I'm trying to read too deeply into it :)

File: packages/cli-repl/src/mongocryptd-manager.ts [hunk 4]
```diff
@@ -116,8 +144,11 @@ export class MongocryptdManager {
     return proc;
   }
 
-  // Try the passed paths to executables for spawning, and parse the process'es
-  // log to understand on what path/port it is listening on.
+  /**
+   * Try to spawn mongocryptd using the paths passed to the constructor,
+   * and parse the process'es log to understand on what path/port it
```
Comment (lerouxb) [inline packages/cli-repl/src/mongocryptd-manager.ts:149]: https://forum.wordreference.com/threads/process-or-processs.1704502/ or https://www.google.com/search?q=What+is+the+possessive+form+of+process%3F&sa=X&ved=2ahUKEwiDi7_w69ryAhVLZMAKHc-7DZoQzmd6BAgOEAU&biw=1676&bih=1336 :)

File: packages/cli-repl/src/smoke-tests.ts [hunk 5]
```diff
@@ -5,9 +5,15 @@ import { once } from 'events';
 import { redactURICredentials } from '@mongosh/history';
 import fleSmokeTestScript from './smoke-tests-fle';
 
-// Run smoke tests on a executable, e.g.
-// runSmokeTests("mongodb://localhost", "/path/to/mongosh.exe") or
-// runSmokeTests(undefined, "/path/to/node", "packages/cli-repl/bin/mongosh.js").
+/**
+ * Run smoke tests on a executable, e.g.
```
Comment (lerouxb) [inline packages/cli-repl/src/smoke-tests.ts:9]: an executable :)

**Review by lerouxb — APPROVED** [review]
Just some very minor nitpicking.

File: packages/cli-repl/src/log-writer.ts (hunk 3 above)
Comment (addaleax) [reply packages/cli-repl/src/log-writer.ts:165]: > Maybe I'm trying to read too deeply into it :)

Yes :) I honestly don’t think it matters, but I’ve tried to at least be mostly consistent for members of the same class/type.
